### PR TITLE
Add an extended prelude for the 2018 editions

### PR DESCRIPTION
**Pull request by SimonSapin**
As proposed in https://github.com/rust-lang/rust/issues/51418.

----

**Update:** Per discussion below this PR also needs to add a warning for 2015 edition code that would break with the new prelude. Help wanted: any ideas how to make that happen?

### CODE REVIEW

**Comment by leodasvacas** [comment]
Yay more prelude. But how will the decision be made for proposed extensions such as: `Debug`, `Display`, `HashMap`, `HashSet`, `FromIterator`.

**Comment by killercup** [comment]
I'm not sure if ~~we~~ rustfix can easily fix errors like https://github.com/carllerche/h2/issues/251 to migrate use to the new prelude.

I think we can, in theory, write lint to the rust-2018-compatibility group that changes calls to methods that become ambiguous to their fully qualified form. But I'd rather hear from @Manishearth who has way more experience with this and can better judge the ~~horrors~~ edge cases we might run into.

**Comment by SimonSapin** [comment]
@leodasvacas I think that a decision for each of these needs to be made on a case-by-case basis, and not necessarily in this PR.

@killercup I just checked https://rust-lang.github.io/rfcs/2052-epochs.html#a-broad-policy-on-epoch-changes and having rustfix automate the transition would be nice to have but per the RFC it is a requirement that code that would break *warns* in the previous edition. Unfortunately I don’t know how to go about implementing that warning…

**Comment by Manishearth** [comment]
Lint feels a bit tricky to handle. Unsure what edge cases exist, though, so it might not be that tricky.

File: src/test/ui/issue-35675.stderr [hunk 1]
```diff
@@ -42,6 +42,7 @@ error[E0573]: expected type, found variant `Ok`
 LL | fn foo() -> Ok {
    |             ^^ not a type
    |
+   = help: there is an enum variant `std::prelude::rust2018::Ok`, try using `std::prelude::rust2018`?
    = help: there is an enum variant `std::prelude::v1::Ok`, try using `std::prelude::v1`?
```
Comment (estebank) [inline src/test/ui/issue-35675.stderr]: Just noticed this terrible suggestion. `std::prelude::v1` is not an enum!

**Comment by pietroalbini** [comment]
Ping from triage! What's the status on this?

**Pull request closed without merging**